tlatools/org.lamport.tlatools/src/tla2sany/explorer/Explorer.java

```diff
@@ -96,12 +97,27 @@ private boolean getLine() {
 		return true;
 	}
 
+	/**
+	 * Print diagnostics about the given module that require access to the
+	 * semantic error log.
+	 *
+	 * @param node The module to print info about.
+	 */
+	private void printModuleNodeDetails(ModuleNode node) {
+		System.out.println("Constant module: " + node.isConstant(this.errors));
+		int errorCount = this.errors.getNumErrors();
+		System.out.print("Module errors: " + (errorCount == 0 ? "none" : Integer.toString(errorCount)));
+	}
+
 	// Integer command
 	private void printNode(int depth) {
 
 		// See if the object requested is already in the table
 		if ((obj = (ExploreNode) semNodesTable.get(icmd)) != null) {
 			// Print tree to depth of icmd2
+			if (obj instanceof ModuleNode) {
+				printModuleNodeDetails((ModuleNode)obj);
+			}
```

Comment:
> instead of the `ModuleNode.toString()` method accessing the static semantic error log to print out some stuff, we move that access into the `Explorer` class itself

An alternative would be to add `Errors` as a parameter to the `toString(depth)` method:

```java
public String toString(int depth, Errors errors) { ... }
```

Although that would result in more lines changed, I think it's cleaner in other ways:
 - SANY's output wouldn't change
 - No awkward `instanceof` checks

Virtually all of the calls to `toString(depth)` are recursive calls, so there are actually very few "real" changes to introduce that parameter.

Reply:
(If you do decide to do it that way, which I endorse, there are exactly two awkward call sites that look like `throw new UnsupportedOperationException("xml export is not yet supported for: " + getClass() + " with toString: " + toString(100));` -- these are dead code, so removing the `toString(100)` call should be fine.)

Reply:
I started with that approach but there are *a lot* of `toString(int)` overloads would have to be changed, in all the various node classes; also `toString(int)` is only ever used by the `Explorer` class and the `Errors` is only ever used by `ModuleNode`.

Reply:
Eclipse `(Refactor > Change Method Signature)` (and probably other IDEs as well) can automate such a refactoring.